src/util/key.rs

```diff
@@ -69,14 +69,36 @@ impl From<secp256k1::Error> for Error {
 }
 
 /// A Bitcoin ECDSA public key
-#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
 pub struct PublicKey {
     /// Whether this public key should be serialized as compressed
     pub compressed: bool,
     /// The actual ECDSA key
     pub key: secp256k1::PublicKey,
 }
 
+impl Ord for PublicKey {
+
+    // implement lexical partial ord for PublicKey as per BIP 67 if both keys are
+    // compressed/uncompressed. Since, compressed Pk's start with 0x02 and 0x03
+    // they are always less than the 0x04 uncompressed keys
+    fn cmp(&self, other: &PublicKey) -> cmp::Ordering {
+        match (self.compressed, other.compressed){
+            (true, false) => cmp::Ordering::Less,
+            (false, true) => cmp::Ordering::Greater,
+            (true, true) => self.key.serialize().cmp(&other.key.serialize()),
```

Comment:
`serialize` is expensive. If someone tries to run sorting on a vector of `PublicKey` s that might be a problem, as these will be run over and over.

That might be a fundamental problem and why `cmp` should not be directly implemented on `PublicKey`

Reply:
Good point. Maybe, what we need a method for sorting instead of `cmp` on `PublicKey`

Reply:
I think we can also add that as an efficient method for sorting. I am not sure whether we should impl `Ord` on `PublicKey` or not. But I do think we should remove the default derive impl because that is incorrect.

Reply:
Seems like https://docs.rs/secp256k1/0.19.0/secp256k1/key/struct.PublicKey.html has `Ord` and under the hood it's all just a `pub struct PublicKey([c_uchar; 64]);` so you should be able to use raw ordering of the bytes, instead of trying to serialize anything first.

Reply:
But, bip67(sortedmulti) requires us to convert that 33-byte representation for lexical comparison which requires seralize. My guess is that 64 bytes in `ffi::PublicKey` have the full x and y co-ordinate.

Reply:
I'm not sure if `rust-bitcoin` should implement default `Ord` from just because some one arbitrary BIP needs it a certain way. Anyone implementing BIP67 might just want to use https://doc.rust-lang.org/std/primitive.slice.html#method.sort_by_cached_key and have all their worries taken care of.

Reply:
One argument for implementing `Ord` this way that we match how it's done in bitcoin core. That being said, I don't have a strong preference for any solution. 
We can also have a method in `ultil/bip67.rs` similar to how we have `util/bip143`. In that case, we should not implement `Ord/PartialOrd` on PublicKey to avoid downstream users sorting incorrectly.

Reply:
Just to be clear: I'm just a frequent lurker here, so I'm just staying my opinion. I agree that not having `Ord` makes sense since there's no one obvious and fast `Ord` to have.

Reply:
@dpc we definitely cannot order based on the underlying bytes in rust-secp, those are opaque and subject to change. The fact that rust-secp has an `Ord` impl based on that is a bug there (and there's a PR to fix it).

The reason for this particular ordering is to match what Core does in `sortedmulti`. I agree we shouldn't take pains to support "arbitrary BIPs" but taking pains to match Bitcoin Core, even in weird behavior, is something that we tend to do.

As for the allocation, there is no reason to do this @sanket1729. We should sort first on compressedness and then on serialization after that.

Reply:
Oh, that's what's done. @dpc I'm not sure what you mean by `serialize` being expensive. I doubt that it's even measurable, it basically just reorders bytes.

Reply:
@apoelstra It was  allocating a new `Vec` https://docs.rs/bitcoin/0.25.2/src/bitcoin/util/psbt/serialize.rs.html#59-63 . AFAIU, Rust people consider every allocation expensive. :D

Reply:
@dpc, this is not calling serialize on `bitcoin::PublicKey`, it is calling it on underlying secp key. 

@apoelstra, the serialize function does do the conversion from 64-byte secp key to 33 byte one. I think that involves minimal crypto operation of checking the sign of y-coordinate. In theory, we do that `log(n)` extra times whenever we call sort operation using the default `PartialOrd`.

On the other hand, having a custom sort function should allow do the seralize only once(instead of `log(n)`) in sorting. I do get that there is some performance benefit to this, but don't have a strong intuition to quantify it.

Reply:
It does two field normalizations, which are 15nsec each on my system, and then checks the least significant bit of y.

Reply:
Oh, OK. No allocation is better, thought it is still not ideal. Judgment call, but seems reasonable.

Reply:
Just make sure there potentially won't be any trouble if `PublicKey` has to change internally or anything like that. The fact that this order is a bit arbitrary and doesn't feel "obvious and natural" still worries me a little.

Reply:
It's unfortunately "obvious and natural" in Bitcoin Core, and it is Core's ordering that we are trying to match.